S100SapCallback_OpenApi_Spec.yml

```
@@ -0,0 +1,151 @@
+openapi: 3.0.1
+info:
+  title: UKHO S100 ERP Facade Service API
+  description: |
+    The UKHO S100 ERP Facade Service API handles callback event received from SAP.
+  contact:
+    name: UKHO
+    email: contact@ukho.co.uk
+  version: 0.0.1
+servers:
+  - url: https://erpfacade-dev-api.azurewebsites.net
+    description: Dev URL
+  
+paths:
+  /s100erpfacade/callback/{correlationId}:
```

without wanting to hold this PR up I have some thoughts - and maybe they are better discussed outside this PR

comments on GET /s100erpfacade/callback/{correlationId}:
- is erpfacade is implied by the base url? although I don't know the production format
- we aren't GETting anything 
- callback is a type of webhook - so should we use the term webhook?
- S100 is part of what we're implementing - but should we be describing the business domain rather than the implementation?

my suggestion here would be the following for the reasons below

POST /v1/webhook/sapactions
{
    "correlationId": "f3284226-90df-496d-ab8e-5a4f7e22000c"
}

- should we start including a version number like v1? This is a standard pattern - maybe a wider discussion
- although webhook is used for reciving events - it can also be considered for an async process. equally we could leave it out
- sapactions is a noun and the POST indicates we're updating something about the SAP Actions we sent. Or it could be sapactionsprocessed which aligns with the other webhook events
- posting correlation in the body aligns more closely with other webhook events

The above is based on my experience of implementing according to wider best practices. But a UKHO API Guidelines document would be useful here. As would a wider discussion and alignment on URI formats.

fyi @vijendr13545 @martyn-fewtrell @jrippington 

I tried to tag Marin Rock-Evans but I can't for some reason.

---

Agree, the path doesn't really need to contain `erpfacade`.  The question about GET vs POST maybe due to limitations within SAP? We'd need to check with the SAP Basis team, or look at what was implemented for ADDS Increment 3.

---

Digging into the past, looks like during the increment 3 work, the call-back from SAP was a POST (including lots of posted data).  It makes more sense that this is indeed a post operation.

We've not included versioning information in the path thus far, and as this is not really a service, the versioning is more around the events being received, so not convinced it would be needed here.

I feel that for clarity of what the path is for, a better name maybe:

```suggestion
  /webhook/sapcallback/s100:
```

---

- Agree, `s100erpfacade` is redundant
- The endpoint is to receive/create a new SAP callback. Shall we have `sapcallback` as the resource for it and use POST verb?
```suggestion
  /sapcallback/s100:
```

---

Now I see where you're coming from @k1500666  with regards to the webhook / sapactions, and that makes some sense, and I agree webhook and callback in there is redundant.

I'm thinking through the discussion about the other events and what URI guidance we can provide. The main argument for keeping the EES webhook as a single endpoint is that it is the EES webhook to receive a cloud event (See <https://cloudevents.io/> for details on the Cloud Events spec). At the HTTP API level, it shouldn't matter which cloud event, that is something that can be determined within the implementation, after all the HTTP aspects (including OPTIONS/POST and auth) have been dealt with. The Controller is dealing with the HTTP aspects, other handlers/code should deal with the business logic of what to do with the specific event.

Also, thinking through "webhook" a little more, a webhook is a subscription process. The publisher (in this case EES) makes a webhook available to all subscribers to subscribe to an event. This matches what we're doing for the EES events. However, it doesn't match what we're doing with SAP, as that isn't a generic Pub/Sub, it really is a callback to the earlier single async call.

So, the long and short, I think I like Vijendra's suggestion, its clear what it is for.

---

Hello, as discussed I have updated event webhook endpoints to
POST /api/v2/webhook
OPTIONS /api/v2/webhook

SAP Callback API Endpoint
POST /api/v2/callback/sap/s100actions/processed